File: src/Azdo.Tui/Components/ThemePicker.cs
using Azdo.Tui.Rendering;
using Azdo.Tui.Runtime;
using StyleSet = Azdo.Tui.Styles.Styles;

namespace Azdo.Tui.Components;

/// <summary>Emitted when a theme is selected (≈ <c>components.ThemeSelectedMsg</c>).</summary>
public sealed record ThemeSelectedMsg(string ThemeName) : IMsg;

/// <summary>A modal for selecting the application theme (≈ <c>components.ThemePicker</c>).</summary>
public sealed class ThemePicker
{
    private const int MinModalWidth = 80;

    private readonly StyleSet _styles;
    private bool _visible;
    private int _width;
    private int _height;
    private readonly IReadOnlyList<string> _availableThemes;
    private readonly string _currentTheme;
    private int _cursor;

    public ThemePicker(StyleSet styles, IReadOnlyList<string> available, string current)
    {
        _styles = styles;
        _availableThemes = available;
        _currentTheme = current;

        for (int i = 0; i < available.Count; i++)
        {
            if (available[i] == current) { _cursor = i; break; }
        }
    }

    public void Show() => _visible = true;
    public void Hide() => _visible = false;
    public bool IsVisible => _visible;
    public void SetSize(int width, int height) { _width = width; _height = height; }
    public int GetCursor() => _cursor;

    public Cmd? Update(IMsg msg)
    {
        if (!_visible) return null;
        if (msg is KeyMsg key)
        {
            switch (key.Key)
            {
                case "esc":
                case "q":
                    _visible = false;
                    return null;
                case "up":
                case "k":
                    if (_cursor > 0) _cursor--;
                    return null;
                case "down":
                case "j":
                    if (_cursor < _availableThemes.Count - 1) _cursor++;
                    return null;
                case "enter":
                    var selected = _availableThemes[_cursor];
                    _visible = false;
                    return Commands.Of(new ThemeSelectedMsg(selected));
            }
        }
        return null;
    }

    public string View()
    {
        if (!_visible) return "";

        var theme = _styles.Theme;

        const string titleText = "Select Theme";
        const string helpTextStr = "↑/↓: navigate • enter: select • esc/q: cancel";

        int maxWidth = MinModalWidth;
        if (titleText.Length > maxWidth) maxWidth = titleText.Length;
        if (helpTextStr.Length > maxWidth) maxWidth = helpTextStr.Length;

        foreach (var themeName in _availableThemes)
        {
            int lineLen = ("> " + themeName + " (current)").Length;
            if (lineLen > maxWidth) maxWidth = lineLen;
        }

        var themeList = new System.Text.StringBuilder();
        for (int i = 0; i < _availableThemes.Count; i++)
        {
            var themeName = _availableThemes[i];
            string cursor = i == _cursor ? ">" : " ";
            string isCurrent = themeName == _currentTheme ? " (current)" : "";
            string line = $"{cursor} {themeName}{isCurrent}";

            line = i == _cursor
                ? Style.New().Foreground(theme.SelectForeground).Background(theme.SelectBackground).Width(maxWidth).Render(line)
                : Style.New().Foreground(theme.Foreground).Background(theme.Background).Width(maxWidth).Render(line);

            themeList.Append(line).Append('\n');
        }

        string title = Style.New()
            .Foreground(theme.Primary)
            .Background(theme.Background)
            .Bold()
            .Width(maxWidth)
            .Render("Select Theme");

        string helpText = Style.New()
            .Foreground(theme.ForegroundMuted)
            .Background(theme.Background)
            .Width(maxWidth)
            .Render(helpTextStr);

        string content = Layout.JoinVertical(HAlign.Left, title, "", themeList.ToString(), helpText);

        var modalStyle = Style.New()
            .WithBorder(Border.Rounded)
            .BorderForeground(theme.Border)
            .Padding(1, 2)
            .Background(theme.Background);

        var modal = modalStyle.Render(content);

        if (_width > 0 && _height > 0)
            modal = Layout.Place(_width, _height, HAlign.Center, VAlign.Center, modal);

        return modal;
    }
}

File: reference/internal/ui/components/themepicker.go
package components

import (
	"fmt"

	"github.com/Elpulgo/azdo/internal/ui/styles"
	"github.com/charmbracelet/bubbles/key"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// ThemeSelectedMsg is sent when a theme is selected
type ThemeSelectedMsg struct {
	ThemeName string
}

// ThemePicker is a modal component for selecting themes
type ThemePicker struct {
	styles          *styles.Styles
	visible         bool
	width           int
	height          int
	availableThemes []string
	currentTheme    string
	cursor          int
}

// NewThemePicker creates a new theme picker
func NewThemePicker(appStyles *styles.Styles, availableThemes []string, currentTheme string) ThemePicker {
	// Find the index of the current theme to set initial cursor position
	cursor := 0
	for i, theme := range availableThemes {
		if theme == currentTheme {
			cursor = i
			break
		}
	}

	return ThemePicker{
		styles:          appStyles,
		visible:         false,
		availableThemes: availableThemes,
		currentTheme:    currentTheme,
		cursor:          cursor,
	}
}

// Show makes the theme picker visible
func (t *ThemePicker) Show() {
	t.visible = true
}

// Hide makes the theme picker invisible
func (t *ThemePicker) Hide() {
	t.visible = false
}

// IsVisible returns whether the theme picker is visible
func (t ThemePicker) IsVisible() bool {
	return t.visible
}

// SetSize sets the dimensions for centering
func (t *ThemePicker) SetSize(width, height int) {
	t.width = width
	t.height = height
}

// GetCursor returns the current cursor position
func (t ThemePicker) GetCursor() int {
	return t.cursor
}

// Update handles messages
func (t ThemePicker) Update(msg tea.Msg) (ThemePicker, tea.Cmd) {
	if !t.visible {
		return t, nil
	}

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch {
		case key.Matches(msg, key.NewBinding(key.WithKeys("esc", "q"))):
			t.visible = false
			return t, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("up", "k"))):
			if t.cursor > 0 {
				t.cursor--
			}
			return t, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("down", "j"))):
			if t.cursor < len(t.availableThemes)-1 {
				t.cursor++
			}
			return t, nil

		case key.Matches(msg, key.NewBinding(key.WithKeys("enter"))):
			selectedTheme := t.availableThemes[t.cursor]
			t.visible = false
			return t, func() tea.Msg {
				return ThemeSelectedMsg{ThemeName: selectedTheme}
			}
		}
	}

	return t, nil
}

// View renders the theme picker
func (t ThemePicker) View() string {
	if !t.visible {
		return ""
	}

	// Build the theme list
	// Calculate maximum line width for consistent backgrounds
	// Include title and help text in width calculation, with minimum width
	titleText := "Select Theme"
	helpTextStr := "↑/↓: navigate • enter: select • esc/q: cancel"

	maxWidth := minModalWidth
	if len(titleText) > maxWidth {
		maxWidth = len(titleText)
	}
	if len(helpTextStr) > maxWidth {
		maxWidth = len(helpTextStr)
	}

	for _, themeName := range t.availableThemes {
		lineLen := len("> " + themeName + " (current)")
		if lineLen > maxWidth {
			maxWidth = lineLen
		}
	}

	var themeList string
	for i, themeName := range t.availableThemes {
		cursor := " "
		if i == t.cursor {
			cursor = ">"
		}

		isCurrent := ""
		if themeName == t.currentTheme {
			isCurrent = " (current)"
		}

		line := fmt.Sprintf("%s %s%s", cursor, themeName, isCurrent)

		if i == t.cursor {
			// Highlight selected item
			line = lipgloss.NewStyle().
				Foreground(t.styles.Theme.GetSelectForeground()).
				Background(t.styles.Theme.GetSelectBackground()).
				Width(maxWidth).
				Render(line)
		} else {
			line = lipgloss.NewStyle().
				Foreground(t.styles.Theme.GetForeground()).
				Background(t.styles.Theme.GetBackground()).
				Width(maxWidth).
				Render(line)
		}

		themeList += line + "\n"
	}

	// Build the modal content
	title := lipgloss.NewStyle().
		Foreground(t.styles.Theme.GetPrimary()).
		Background(t.styles.Theme.GetBackground()).
		Bold(true).
		Width(maxWidth).
		Render("Select Theme")

	helpText := lipgloss.NewStyle().
		Foreground(t.styles.Theme.GetForegroundMuted()).
		Background(t.styles.Theme.GetBackground()).
		Width(maxWidth).
		Render("↑/↓: navigate • enter: select • esc/q: cancel")

	content := lipgloss.JoinVertical(
		lipgloss.Left,
		title,
		"",
		themeList,
		helpText,
	)

	// Create modal box
	modalStyle := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		BorderForeground(t.styles.Theme.GetBorder()).
		Padding(1, 2).
		Background(t.styles.Theme.GetBackground())

	modal := modalStyle.Render(content)

	// Center the modal
	if t.width > 0 && t.height > 0 {
		modal = lipgloss.Place(
			t.width,
			t.height,
			lipgloss.Center,
			lipgloss.Center,
			modal,
		)
	}

	return modal
}
